stdlib/immutableset.gr

```diff
@@ -0,0 +1,499 @@
+ *
+ * @param set: The set to inspect
+ * @returns `true` if the given set is empty or `false` otherwise
+ * 
+ * @since v0.5.4
+ */
+export let isEmpty = set => {
+  match (set) {
+    Empty => true,
+    Tree(_) => false,
+  }
+}
+
+let unwrapTree = node => {
+  match (node) {
+    Empty => fail "Impossible: ImmutableSet unwrapTree got an empty tree node",
+    Tree(tree) => tree,
+  }
+}
+
+// helper function for creating a tree node with correct size from
+// two balanced trees
+let makeNode = (key, left, right) => {
+  Tree({ key, size: 1 + size(left) + size(right), left, right })
+}
+
+// note: see Figure 1 of paper referenced above for visual illustration of
+// the rotations below
+
+// node rotation moving the left subtree of the right node to the left side
+let singleL = (key, left, right) => {
+  let { key: rKey, left: rl, right: rr, _ } = unwrapTree(right)
+  makeNode(rKey, makeNode(key, left, rl), rr)
+}
+
+// node rotation moving left child of right tree to the root
+let doubleL = (key, left, right) => {
+  let { key: rKey, left: rl, right: rr, _ } = unwrapTree(right)
+  let { key: rlKey, left: rll, right: rlr, _ } = unwrapTree(rl)
+  makeNode(rlKey, makeNode(key, left, rll), makeNode(rKey, rlr, rr))
+}
+
+// node rotation moving the right subtree of the left node to the right side
+let singleR = (key, left, right) => {
+  let { key: lKey, left: ll, right: lr, _ } = unwrapTree(left)
+  makeNode(lKey, ll, makeNode(key, lr, right))
+}
+
+// node rotation moving right child of left tree to the root
+let doubleR = (key, left, right) => {
+  let { key: lKey, left: ll, right: lr, _ } = unwrapTree(left)
+  let { key: lrKey, left: lrl, right: lrr, _ } = unwrapTree(lr)
+  makeNode(lrKey, makeNode(lKey, ll, lrl), makeNode(key, lrr, right))
+}
+
+// creates a new node after either the left or right trees have just had an
+// element inserted or removed from them, maintaining balance in the tree
+let balancedNode = (key, left, right) => {
+  let makeNodeFn = if (size(left) + size(right) < 2) {
+    makeNode
+  } else if (size(right) > weight * size(left)) {
+    // if the right tree is too much larger than the left then move part of
+    // the right tree to the left side
+    let Tree({ left: rl, right: rr, _ }) = right
+    if (size(rl) < size(rr)) singleL else doubleL
+  } else if (size(left) > weight * size(right)) {
+    // if the left tree is too much larger than the right then move part of
+    // the left tree to the right side
+    let Tree({ left: ll, right: lr, _ }) = left
+    if (size(lr) < size(ll)) singleR else doubleR
+  } else {
+    // if neither tree is too much larger than the other then simply create
+    // a new node
+    makeNode
+  }
+
+  makeNodeFn(key, left, right)
+}
+
+/**
+ * Produces a new set by inserting the given value into the set. If the value
+ * already exists, the new set will have the same elements as the input set.
+ *
+ * @param key: The value to add
+ * @param set: The set to add to
```

Comment:
I don't like "add to" as the end of this sentence. It's better than "modify" like in ImmutableMap but we can do better.